File: 2017/day25/day25.py
import time
steps = 6
rules = {"A": [(1,1,"B"),(0,-1,"B")],
         "B": [(1,-1,"A"),(1,1,"A")]}
tape = {}
init_on = []
init_state = "A"
init_pos = 0

def turing(ones, state, pos):
    cur_val = 1 if pos in ones else 0
    (new_val, dx, new_state) = rules[state][cur_val]
    if new_val == 1:
        ones += [pos]
    else:
        ones = [x for x in ones if x!=pos]
    return ones, new_state, (pos+dx)
    
def turing2(tape, state, pos):
    cur_val = tape[pos] if pos in tape.keys() else 0
    (new_val, dx, new_state) = rules[state][cur_val]
    tape[pos] = new_val
    return tape, new_state, (pos+dx)

def part_one():
    global init_on, init_state, init_pos, steps
    for i in range(steps):
        init_on, init_state, init_pos = turing(init_on, init_state, init_pos)
    print(len(init_on))

def part_one2():
    global tape, init_state, init_pos, steps
    for i in range(steps):
        tape, init_state, init_pos = turing2(tape, init_state, init_pos)
    print(sum(tape.values()))

s = time.time()
part_one2()
print(time.time()-s)

File: 2017/day25/day25.hs
import System.IO
import qualified Data.Map as Map

nine_chunks :: Eq a => [a] -> [[a]]
nine_chunks nums
    |(length nums `mod` 9)/=0 = error "Something has gone horribly wrong"
    |nums == [] = []
    |(length nums == 9) = [nums]
    |otherwise = [(take 9 nums)]++(nine_chunks $ drop 9 nums)

--chunk looks like [cur_state, if 0, to_write, dir, next_state, if 1, to_write, dir, next_state]
parse_chunk :: Num b => [String] -> (String, [(Int, b, String)])
parse_chunk chunk = (chunk!!0, [(r (chunk!!2), d (chunk!!3), chunk!!4), (r (chunk!!6), d (chunk!!7), chunk!!8)])
    where
        r = (read::String->Int)
        d "right" = 1
        d "left" = -1

turing :: (Ord b1, Ord c, Num c) => Map.Map b1 [(Int, c, b2)] -> (Map.Map c Int, b1, c) -> (Map.Map c Int, b2, c)
turing rules (tape, state, pos) = (new_tape, new_state, new_pos)
    where
        to_do = (Map.!) rules state
        current_val = Map.findWithDefault 0 pos tape
        (new_val, dir, new_state) = to_do!!current_val
        !new_tape = Map.insert pos new_val tape
        !new_pos = pos+dir

main = do
    input <- lines <$> readFile "day25.txt"
    let
        init_state = (init.last.words) (input!!0)
        steps = ((read::String->Int).head.(drop 1).reverse.words) (input!!1)
        rules = map (init.last.words) $ filter (/="") (drop 2 input)
        parsed_rules = Map.fromList (map parse_chunk $ nine_chunks rules)
        part_one = (\(a,_,_)->sum $ Map.elems a) $ (iterate (turing parsed_rules) (Map.fromList [], init_state, 0))!!(steps)
    print(part_one)